validator/src/main.rs

```diff
@@ -0,0 +1,248 @@
+/// redirect connections to a Linkerd proxy (with an outbound port of
+/// `listen_addr`).
+async fn validate(listen_addr: SocketAddr, connect_addr: SocketAddr) -> Result<()> {
+    // First, bind the server address so that all connections can be processed
+    // by the server.
+    let listener = TcpListener::bind(listen_addr).await?;
+    info!("Listening for connections on {listen_addr}");
+
+    // Generate a random token to be sent from the server. Clients use the
+    // server response to ensure that it is connecting to this process.
+    let token = format!(
+        "{}\n",
+        Alphanumeric.sample_string(&mut rand::thread_rng(), 63)
+    );
+    debug!(?token);
+    let token = Bytes::from(token);
+
+    // Spawn a server on a background task that writes the response and then
+    // closes client the connection.
+    tokio::spawn(serve(listener, token.clone()).in_current_span());
+
+    // Connect to an arbitrary address, read data from the connection, and fail
+    // if it doesn't match the server's token.
+    info!("Connecting to {connect_addr}");
+    let data = connect(connect_addr, token.len()).await?;
+    debug!(data = ?String::from_utf8_lossy(&*data), size = data.len());
+    ensure!(
+        data == token,
+        "expected client to receive {:?}; got {:?} instead",
+        String::from_utf8_lossy(&*token),
+        String::from_utf8_lossy(&*data),
+    );
+    Ok(())
+}
+
+/// Accepts connections from `listener`, writes `token` to the socket, and then closes the
+/// connection.
+#[tracing::instrument(level = "debug", skip_all)]
+async fn serve(listener: TcpListener, token: Bytes) {
+    loop {
+        let (mut socket, client_addr) = listener
+            .accept()
+            .await
+            .expect("Failed to establish connection");
+        let token = token.clone();
+        tokio::spawn(
+            async move {
+                debug!("Accepted");
+                // We expect this write to complete instantaneously, so a timeout is not needed
+                // here.
+                match socket.write_all(&*token).await {
+                    Ok(()) => debug!(bytes = token.len(), "Wrote message to client"),
+                    Err(error) => error!(%error, "Failed to write bytes to client"),
+                }
+            }
+            .instrument(tracing::info_span!("conn", client.addr = %client_addr)),
+        );
+    }
+}
+
+/// Connects to the target address and reads exactly `size` bytes.
+#[tracing::instrument(level = "debug", skip_all)]
+async fn connect(addr: SocketAddr, size: usize) -> Result<Bytes> {
+    let mut socket = TcpStream::connect(addr).await?;
+    debug!(client.addr = %socket.local_addr()?, "Connected");
+
+    socket
+        .readable()
+        .await
+        .context("cannot read from client socket")?;
+
+    let mut buf = BytesMut::with_capacity(size);
+    while buf.len() != size {
+        let size = socket.read_buf(&mut buf).await?;
```

Comment:
reposting question here since I closed the other PR:

> Will shadowing size here have an effect on the outer loop condition? Could this like finish the loop in one iteration if for example we only partially fill the buffer in the first read?

Probably not an issue here, I assume the scope inside the while loop will be different so it won't impact the condition.

Reply:
That's right. We could change the outer `size` to `capacity` or we could change the inner size to `sz` (since it's only used in a small scope -- small scope => small name); but this would only be for readability. It's correct as-is.